lib/Runtime/Executor/ThreadPoolExecutor.cpp

```diff
@@ -0,0 +1,377 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ThreadPoolExecutor.h"
+
+#include "glow/Backends/DeviceManager.h"
+#include "glow/Graph/Context.h"
+
+#include <queue>
+#include <unordered_set>
+
+namespace glow {
+namespace runtime {
+
+ExecutionState::ExecutionState(RunIdentifierTy id, const DAGNode *root,
+                               DoneCbTy doneCb)
+    : cb_(doneCb), resultCtx_(std::make_unique<Context>()), inflightNodes_(0),
+      resultCode_(ResultCode::Ready) {
+  // Create a queue for the breadth-first traversal through the graph.
+  std::queue<const DAGNode *> bfsQueue;
+  // Create a set to keep track of visited nodes.
+  std::unordered_set<const DAGNode *> visited;
+
+  // Place the root nodes in the queue.
+  for (const auto &node : root->children) {
+    bfsQueue.push(node);
+  }
+
+  // Breadth-first search.
+  while (!bfsQueue.empty()) {
+    // Get the next node in the BFS queue.
+    const DAGNode *node = bfsQueue.front();
+    bfsQueue.pop();
+
+    // Mark the node as visited.
+    visited.insert(node);
+
+    // Make a counter for the number of node parents done.
+    nodeParentsDone_[node] = 0;
+
+    // Make an (empty) input Context for the node.
+    inputCtxs_.insert(std::make_pair(node, std::make_unique<Context>()));
+
+    // Push all unvisited children onto the BFS queue.
+    for (const auto &child : node->children) {
+      if (!visited.count(child)) {
+        bfsQueue.push(child);
+      }
+    }
+  }
+}
+
+void ExecutionState::insertIntoNodeCtx(const DAGNode *node, Placeholder *P,
+                                       Tensor &&T) {
+  // Get a raw pointer to the input Context for the node. It should have
+  // been created in the constructor.
+  auto ctxIt = inputCtxs_.find(node);
+
+  if (ctxIt == inputCtxs_.end()) {
+    assert(false && "Input context not found but should exist!");
+  }
+
+  Context *ctx = (ctxIt->second).get();
+  assert(ctx && "Input context for node is null");
+
+  // Insert the placeholder-tensor pair.
+  std::lock_guard<std::mutex> lock(contextMtx_);
```

Comment:
Only one thread can write to any context at a time? Is there performance to be gained by having a lock per Context?

Reply:
Yes, only one thread can write to any `Context` at a time, since writing to it might involve inserting new key-value pairs into the underlying map inside the `Context`. This can happen if a node has multiple parent nodes that feed their outputs to it.

Reply:
I think there might be some performance to be gained but I cannot say for certain without measurement. I'll keep it mind for when do start benchmarking and looking for optimization opportunities.